### ISSUE TITLE
Between Dates filter behaves inconsistently based on whether the column is from a joined table or not

### ISSUE THREAD

**Issue by scottbradybuckii** [metabase/metabase#12872]
**Describe the bug**
When adding a date filter between two dates (I am currently using Postgres), it will make the filter inclusive of the end date if the column is from the base table, but it will make it exclusive of the end date if the column is from a joined table.

**Logs**
No errors are generated

**To Reproduce**
1. Add a date filter to a column on the base table of a Question that is between two dates
2. Click on the "View the SQL" button for that Question
3. Verify that the dates are inclusive - ie. if your filter is between 2020-07-06 -> 2020-07-08, then the query would looks something like: 
`WHERE ("public"."line_items"."created_at" >= timestamp with time zone '2020-07-06 00:00:00.000Z'
   AND "public"."line_items"."created_at" < timestamp with time zone '2020-07-09 00:00:00.000Z')`
4. Add a join to the Question
5. Change the same date filter to filter down based on a date column in the joined table
6. Click on the "View the SQL" button for that Question
7. Verify that the dates are now exclusive:
`WHERE "Transaction"."created_at" BETWEEN timestamp with time zone '2020-07-06 00:00:00.000Z'
   AND timestamp with time zone '2020-07-08 00:00:00.000Z'`

**Expected behavior**
The inclusivity and exclusivity shouldn't change based on whether you are filtering a column from a base table or joined table

**Screenshots**
Inclusive from base table:
![Screenshot from 2020-07-08 15-02-45](https://user-images.githubusercontent.com/60900548/86968813-c70cf300-c13a-11ea-86e0-f9b9ff831f16.png)
![Screenshot from 2020-07-08 16-45-44](https://user-images.githubusercontent.com/60900548/86968818-c83e2000-c13a-11ea-8af8-a9f84908ce1a.png)

Exclusive from joined table:
![Screenshot from 2020-07-08 15-02-57](https://user-images.githubusercontent.com/60900548/86968816-c7a58980-c13a-11ea-8657-e8ea5cd63893.png)
![Screenshot from 2020-07-08 16-46-29](https://user-images.githubusercontent.com/60900548/86968819-c83e2000-c13a-11ea-99ab-4a696b94aa4f.png)

**Information about your Metabase Installation:**
- Your browser and the version: Chromium 81.0.4044.138
- Your operating system: Manjaro
- Your databases: Postgres
- Metabase version: v0.35.4
- Metabase hosting environment: Docker running on Ubuntu 16.04
- Metabase internal database: H2

**Severity**
Moderately severe. We have been using Metabase for over a year now and just noticed it, but it is worrisome because we are putting together weekly financial reports and this off-by-one day error has messed up our reporting. We can train users to add an extra day to reports that are filtering based on a joined date, but if this bug gets fixed, then those reports will be off by a day.

**Comment by flamber** [comment]
I have a feeling that it's related to #10824.
Can reproduce with H2 application database:
```
SELECT "PUBLIC"."CORE_USER"."LAST_LOGIN" AS "LAST_LOGIN", "Core User"."LAST_LOGIN" AS "LAST_LOGIN_2"
FROM "PUBLIC"."CORE_USER"
LEFT JOIN "PUBLIC"."CORE_USER" "Core User" ON "PUBLIC"."CORE_USER"."ID" = "Core User"."ID"
WHERE (
  ("PUBLIC"."CORE_USER"."LAST_LOGIN" >= timestamp with time zone '2020-01-01 00:00:00.000+01:00'
    AND "PUBLIC"."CORE_USER"."LAST_LOGIN" < timestamp with time zone '2020-07-10 00:00:00.000+02:00')
  AND "Core User"."LAST_LOGIN" BETWEEN timestamp with time zone '2020-01-01 00:00:00.000+01:00' AND timestamp with time zone '2020-07-09 00:00:00.000+02:00'
)
LIMIT 1048576
```
![image](https://user-images.githubusercontent.com/1447303/87027687-d42eee00-c1dd-11ea-8640-3a345587bda2.png)

**Comment by nemanjaglumac** [comment]
@camsaul since you're assigned to work on this issue, just a note that a repro has been made, and it should be unskipped and merged together with a fix.

**Comment by camsaul** [comment]
Fixed by #13991
